File: Clone-Graph/Clone-Graph.cpp

/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> ans;
        if(node==NULL)
            return node;
        queue<UndirectedGraphNode*> q;
        q.push(node);
        while(!q.empty())
        {
            UndirectedGraphNode *s = q.front();
            q.pop();
            if(!ans.count(s))
            {
                UndirectedGraphNode *next = new UndirectedGraphNode(s->label);
                ans[s]=next;
            }
            for(int i=0;i<s->neighbors.size();i++)
            {
                UndirectedGraphNode *next = s->neighbors[i];
                if(!ans.count(next))
                {
                    UndirectedGraphNode *p = new UndirectedGraphNode(next->label);
                    ans[next]=p;
                    q.push(next);
                }
                ans[s]->neighbors.push_back(ans[next]);
            }
        }
        return ans[node];
    }
};


 

File: solutions/Clone-Graph/Clone-Graph.py


# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        dict = {}
        if node==None:
            return node
        q=[]
        q.append(node)
        while len(q)>0:
            s = q[0]
            q.pop(0)
            if s not in dict:
                next = UndirectedGraphNode(s.label)
                dict[s] = next
            for i in range(len(s.neighbors)):
                next = s.neighbors[i]
                if next not in dict:
                    p = UndirectedGraphNode(next.label)
                    dict[next]=p
                    q.append(next)
                dict[s].neighbors.append(dict[next])
        return dict[node]
 